stumpy/mmotifs.py

```diff
@@ -0,0 +1,186 @@
+        finding the dimension of each motif
+    """
+    T = core._preprocess(T)
+
+    m = T.shape[-1] - P.shape[-1] + 1
+
+    if max_motifs < 1:  # pragma: no cover
+        logger.warning(
+            "The maximum number of motifs, `max_motifs`, "
+            "must be greater than or equal to 1"
+        )
+        logger.warning("`max_motifs` has been set to `1`")
+        max_motifs = 1
+
+    T, M_T, Σ_T = core.preprocess(T, m)
+
+    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
+
+    if max_matches is None:
+        max_matches = np.inf
+
+    if cutoffs is None:
+        P_copy = P.copy().astype(np.float64)
+        P_copy[np.isinf(P_copy)] = np.nan
+        cutoffs = np.nanmax(
+            [
+                np.nanmean(P_copy - 2.0 * np.nanstd(P_copy, axis=0), axis=1),
+                np.nanmin(P_copy, axis=1),
+            ],
+            axis=0,
+        )
+    elif isinstance(cutoffs, int) or isinstance(cutoffs, float):
+        cutoffs = np.full(T.shape[0], cutoffs)
+
+    motif_distances = []
+    motif_indices = []
+    motif_subspaces = []
+    motif_mdls = []
+
+    candidate_idx = np.argmin(P, axis=1)
+    nn_idx = I[np.arange(len(candidate_idx)), candidate_idx]
+
+    while len(motif_distances) < max_motifs:
+        mdls, subspaces = mdl(T, m, candidate_idx, nn_idx)
+        k = np.argmin(mdls)
+        motif_mdls.append(mdls)
+
+        motif_idx = candidate_idx[k]
+        motif_value = P[k, motif_idx]
+
+        if (
+            motif_value > cutoffs[k]
+            or not np.isfinite(motif_value)
+            or (isinstance(max_distance, float) and motif_value > max_distance)
+        ):
+            break
+
+        motif_subspaces.append(subspaces[k])
+
+        # if isinstance(max_distance, float) and motif_value > max_distance:
+        #     break
+
+        query_matches = match(
+            Q=T[subspaces[k], motif_idx : motif_idx + m],
+            T=T[subspaces[k]],
+            M_T=M_T[subspaces[k]],
+            Σ_T=Σ_T[subspaces[k]],
+            max_matches=max_matches,
+            max_distance=max_distance,
+            atol=atol,
+            # normalize=normalize,
+            # p=p
+        )
+
+        if len(query_matches) > min_neighbors:
+            motif_distances.append(query_matches[:, 0])
+            motif_indices.append(query_matches[:, 1])
+
+        for idx in query_matches[:, 1]:
+            core.apply_exclusion_zone(P, idx, excl_zone, np.inf)
+        candidate_idx = np.argmin(P, axis=1)
+        nn_idx = I[np.arange(len(candidate_idx)), candidate_idx]
+
+    motif_distances = core._jagged_list_to_array(
+        motif_distances, fill_value=np.nan, dtype=np.float64
+    )
+    motif_indices = core._jagged_list_to_array(
+        motif_indices, fill_value=-1, dtype=np.int64
+    )
+
+    return (
+        motif_distances,
+        motif_indices,
+        np.array(motif_subspaces, dtype=object),

```

Comment:
I don't think the `dtype` should be `object` since they should all have the same data type within `motif_subspaces` and `motif_mdls`, respectively.

I'm guessing that the problem that you are encountering is that the length of each row is different. Thus, we should avoid passing back a NumPy array and simply return back the ragged/jagged list of numpy arrays instead (and update the return types in our docstring):

```
return motif_distances, motif_indices, motif_subspaces, motif_mdls
```

Reply:
So then all four return values should be lists?
So you want me to remove
```
motif_distances = core._jagged_list_to_array(
        motif_distances, fill_value=np.nan, dtype=np.float64
    )
 motif_indices = core._jagged_list_to_array(
        motif_indices, fill_value=-1, dtype=np.int64
    )
```
as well?

Reply:
So, I wouldn't change `motif_distances` or `motif_indices`.  

Let me tell you where my brain is at regarding `motif_subspaces` and `motif_mdls`. Usually, for "jagged/ragged" data structures, I always leave them a list of `numpy` arrays as the `list` type implies something that isn't perfectly rectangular. This way of thinking is pretty common for more `numpy` users and so I wanted to be consistent. 

I don't know how much value there is to converting `motif_subspaces` and `motif_mdls` into a rectangular shape because all of the places that we will need to pad with extra values would need to be obvious that they are not to be used/considered. Perhaps, we could do the same thing and do:

```
motif_mdls = core._jagged_list_to_array(
        motif_mdls, fill_value=np.nan, dtype=np.float64
    )
 motif_subspaces = core._jagged_list_to_array(
        motif_subspaces, fill_value=-1, dtype=np.int64
    )
```

However, then you'd have to be careful that `-1` isn't chosen as a subspace (which, coincidentally, is the last index value of a numpy array). I am happy either way (i.e., to leave `motif_mdls` and `motif_subspaces` as lists or rectangular arrays) but I would want to avoid non-rectangular numpy arrays. Most importantly, as a user, you do have a say in this. What are your thoughts?

Reply:
My personal preference are lists, but I think that depends on the person, you are asking.

> Let me tell you where my brain is at regarding motif_subspaces and motif_mdls. Usually, for "jagged/ragged" data structures, I always leave them a list of numpy arrays as the list type implies something that isn't perfectly rectangular. This way of thinking is pretty common for more numpy users and so I wanted to be consistent.

Ahh, you're right! I haven't thought about that yet. But this also means, that I have to consider that point in `motif_distances` and `motif_indices` since the motifs can have a different number of matches depending on the `max_distance` parameter.

Reply:
Indeed! You don't want to get burned by the fact that in:

```
motif_indices = [[1, 2, 3], [9, -1, -1]]
```

The first motif has 3 legitimate matches but the second only has 1 legitimate match! But if you were to blindly apply the the indices then you would be selecting `T[-1 : -1 + m]` which makes no sense. Unfortunately, in `numpy` there is no reasonable way to represent a "bad" integer value (i.e., you can't assign `np.nan` or `np.inf` to an integer array - at least, not that I'm aware of). This stuff is tricky and we are trying the best that we can :(